### PR TITLE
[CS-2660] Introduce generations to realm reindexing

### PR DESCRIPTION
**Pull request by mattmcmanus**

### CODE REVIEW

**Review by habdelra — APPROVED** [review]
LGTM

File: packages/hub/services/search-index.ts [hunk 1]
```diff
@@ -13,9 +13,13 @@ export class SearchIndex {
   private database = inject('database-manager', { as: 'database' });
 
   async indexAllRealms(): Promise<void> {
-    await this.runIndexing(async (ops) => {
-      await Promise.all(this.realmManager.realms.map((realm) => realm.reindex(ops, undefined)));
-    });
+    await Promise.all(
+      this.realmManager.realms.map((realm) => {
+        return this.runIndexing(realm.url, async (ops) => {
+          realm.reindex(ops, undefined);
```
Comment (ef4) [inline packages/hub/services/search-index.ts:19]: This looks like an un-awaited promise?

File: packages/hub/services/search-index.ts [hunk 2]
```diff
@@ -75,7 +81,9 @@ export interface IndexerHandle {
 }
 
 class IndexerRun implements IndexerHandle {
-  constructor(private db: PoolClient, private builder: CardBuilder) {}
+  generation?: number;
```
Comment (ef4) [inline packages/hub/services/search-index.ts:84]: I think this should be private.

File: packages/hub/services/search-index.ts [hunk 3]
```diff
@@ -99,9 +109,13 @@ class IndexerRun implements IndexerHandle {
...
   }
 
-  async beginReplaceAll(): Promise<void> {}
+  async beginReplaceAll(): Promise<void> {
+    this.generation = Math.floor(Math.random() * 1000000000);
+  }
 
-  async finishReplaceAll(): Promise<void> {}
+  async finishReplaceAll(): Promise<void> {
+    this.db.query('DELETE FROM cards where realm = $1 AND generation != $2', [this.realmURL, this.generation]);
```
Comment (ef4) [inline packages/hub/services/search-index.ts:117]: Un-awaited promise?

File: packages/hub/services/search-index.ts [hunk 4]
```diff
@@ -87,6 +95,8 @@ class IndexerRun implements IndexerHandle {
     let compiledCard = await this.builder.compileCardFromRaw(card);
     let expression = upsert('cards', 'cards_pkey', {
       url: param(card.url),
+      realm: param(this.realmURL || null),
```
Comment (ef4) [inline packages/hub/services/search-index.ts:98]: The db migration said this column is `NOT NULL`.

**Review by ef4 — CHANGES_REQUESTED** [review]

**Pull request merged**